### PR TITLE
Fix GCM calculation with very long IV

### LINKED ISSUE #4884: wrong gcm computation when iv length is big enough in main branch
### Summary
In the GCM cipher mode, it may be wrong in the computation of Y<sub>0</sub> when the bit length of iv is not smaller than 2<sup>32</sup>.

### System information

Mbed TLS version (number or commit id): 
Operating system and version: 
Configuration (if not default, please attach `mbedtls_config.h`): 
Compiler and options (if you used a pre-built binary, please indicate how you obtained it): 
Additional environment information: 

### Expected behavior
_X<sub>i</sub> = (X<sub>m+n</sub>) ^ (len(A) || len(C)) . H_

Since the bit length of _C_(here is iv) can be upper to 2<sup>64</sup>. So, it may be something like

`PUT_UINT32_BE( iv_len * 8, work_buf, 8 )`

### Actual behavior
`PUT_UINT32_BE( iv_len * 8, work_buf, 12 )`

### PR DESCRIPTION
**Pull request by openluopworld**
Fixes #4884 
An initialization vector IV can have any number of bits between 1 and
2^64. So it should be filled to the lower 64-bit in the last step
when computing ghash.

Notes:
* Pull requests cannot be accepted until the PR follows the [contributing guidelines](../CONTRIBUTING.md). In particular, each commit must have at least one `Signed-off-by:` line from the committer to certify that the contribution is made under the terms of the [Developer Certificate of Origin](../dco.txt).
* This is just a template, so feel free to use/remove the unnecessary things
## Description
A few sentences describing the overall goals of the pull request's commits.

## Status
**READY/IN DEVELOPMENT/HOLD**

## Requires Backporting
Yes: https://github.com/ARMmbed/mbedtls/pull/5132

## Migrations
If there is any API change, what's the incentive and logic for it.

YES | NO

## Additional comments
Any additional information that could be of interest

## Todos
- [ ] Tests
- [ ] Documentation
- [x] Changelog updated
- [x] Backported

## Steps to test or reproduce
Outline the steps to test or reproduce the PR here.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Thanks for the fix! To confirm that we have the right to use your code, please add a signoff line to each commit message idicating that you accept the [Developer Certificate of Origin](https://github.com/ARMmbed/mbedtls/dco.txt). The “DCO” check needs to pass.

File: library/gcm.c
Comment (davidhorstmann-arm) [inline library/gcm.c:282]: 2 small nits:
- We have a macro: `MBEDTLS_PUT_UINT64_BE()` - please use this instead as it is neater than using 2 calls to `MBEDTLS_PUT_UINT32_BE()`
- `* 8` is more readable than `<< 3`, as the intent of the code is clearer (and most compilers will replace the multiply with a shift anyway).

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
A couple of nits, but looks good otherwise!

**Comment by gilles-peskine-arm** [comment]
The CI is still failing. Please fix the issues reported by Travis. Then, if there are remaining issues on Jenkins, which unfortunately isn't publicly accessible, one of us maintainers can post the relevant logs.

**Comment by davidhorstmann-arm** [comment]
I believe CI is failing because `iv_len` is a `size_t` which may be 32-bit on some platforms. Casting it to a `uint64_t` should solve this issue.

This raises the question of whether we should cast to `uint64_t` within the `MBEDTLS_PUT_UINT64` macro itself to prevent bad things happening in future.

File: library/gcm.c
Comment (davidhorstmann-arm) [inline library/gcm.c:281]: Nit: extra space before `work_buf`
```suggestion
        MBEDTLS_PUT_UINT64_BE( iv_len * 8, work_buf, 8 );
```

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
One nit as well as making the CI pass (as I mentioned, probably casting `iv_len` to a `uint64_t` will solve it)

File: library/gcm.c
Comment (openluopworld) [reply library/gcm.c:281]: thanks

**Comment by gilles-peskine-arm** [comment]
Please remember to include a signoff line in every commit message. Otherwise we aren't allowed to use your code.

https://stackoverflow.com/questions/13043357/git-sign-off-previous-commits and https://stackoverflow.com/questions/10161198/is-there-a-way-to-autosign-commits-in-git-with-a-gpg-key may help.

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
Looks fine to me apart from the commit signoffs.

**Review by AndrzejKurek — APPROVED** [review]
Looks good. It would be also great to have a test case exposing the case of y0 miscalculation, but it might be also added separately, if at all.

**Comment by gilles-peskine-arm** [comment]
@AndrzejKurek The incorrect case is for an IV length over 2^32, which is not practical for us to test (since our tests should be able to run on a microcontroller and it can't process 4GB in a reasonable time). We could add some separate large-data tests that don't run everywhere but that would be outside the scope here.

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM modulo changelog etc

**Comment by gilles-peskine-arm** [comment]
Can you please force-push to remove the merge commit and go back to 151ccb297766b6263cdce69f8ca0b785b5323499? A merge commit gives us extra review work and makes the history more complicated. Please don't include merge commits in a pull request unless the history has diverged and the merge is required to avoid a conflict.

**Comment by mpg** [comment]
> Can you please force-push to remove the merge commit and go back to 151ccb2? 

@openluopworld This PR and its backport are ready for merge except for this, can you move this PR back to 151ccb2 in order to unblock them? Thanks!

**Comment by tom-cosgrove-arm** [comment]
@mpg Following lack of response from originator, I have raised PR #5213 which is just this PR rolled back to before the merge commit, as requested

**Comment by gilles-peskine-arm** [comment]
Continued in https://github.com/ARMmbed/mbedtls/pull/5213 without the spurious merge commit.

**Pull request closed without merging**
